src/runtime/virtcontainers/qemu_amd64.go

```diff
@@ -279,11 +335,281 @@ func (q *qemuAmd64) appendProtectionDevice(devices []govmmQemu.Device, firmware,
+
+// GUID: 44baf731-3a2f-4bd7-9af1-41e29169781d
+var sevInitrdEntryGuid = guidLE{0x31, 0xf7, 0xba, 0x44, 0x2f, 0x3a, 0xd7, 0x4b, 0x9a, 0xf1, 0x41, 0xe2, 0x91, 0x69, 0x78, 0x1d}
+
+// GUID: 97d02dd8-bd20-4c94-aa78-e7714d36ab2a
+var sevCmdlineEntryGuid = guidLE{0xd8, 0x2d, 0xd0, 0x97, 0x20, 0xbd, 0x94, 0x4c, 0xaa, 0x78, 0xe7, 0x71, 0x4d, 0x36, 0xab, 0x2a}
+
+type sevHashTableEntry struct {
+	entryGuid guidLE
+	length    uint16
+	hash      [sha256.Size]byte
+}
+
+type sevHashTable struct {
+	tableGuid guidLE
+	length    uint16
+	cmdline   sevHashTableEntry
+	initrd    sevHashTableEntry
+	kernel    sevHashTableEntry
+}
+
+type paddedSevHashTable struct {
+	table   sevHashTable
+	padding [8]byte
+}
+
+func fileSha256(filename string) (res [sha256.Size]byte, err error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return res, err
+	}
+	defer f.Close()
+
+	digest := sha256.New()
+	if _, err := io.Copy(digest, f); err != nil {
+		return res, err
+	}
+
+	copy(res[:], digest.Sum(nil))
+	return res, nil
+}
+
+func constructSevHashesTable(kernelPath, initrdPath, cmdline string) ([]byte, error) {
+	kernelHash, err := fileSha256(kernelPath)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	initrdHash, err := fileSha256(initrdPath)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	cmdlineHash := sha256.Sum256(append([]byte(cmdline), 0))
+
+	buf := new(bytes.Buffer)
+	err = binary.Write(buf, binary.LittleEndian, sevHashTableEntry{})
+	if err != nil {
+		return []byte{}, err
+	}
+	entrySize := uint16(buf.Len())
+
+	buf = new(bytes.Buffer)
+	err = binary.Write(buf, binary.LittleEndian, sevHashTable{})
+	if err != nil {
+		return []byte{}, err
+	}
+	tableSize := uint16(buf.Len())
+
+	ht := paddedSevHashTable{
+		table: sevHashTable{
+			tableGuid: sevHashTableHeaderGuid,
+			length:    tableSize,
+			cmdline: sevHashTableEntry{
+				entryGuid: sevCmdlineEntryGuid,
+				length:    entrySize,
+				hash:      cmdlineHash,
+			},
+			initrd: sevHashTableEntry{
+				entryGuid: sevInitrdEntryGuid,
+				length:    entrySize,
+				hash:      initrdHash,
+			},
+			kernel: sevHashTableEntry{
+				entryGuid: sevKernelEntryGuid,
+				length:    entrySize,
+				hash:      kernelHash,
+			},
+		},
+		padding: [8]byte{0, 0, 0, 0, 0, 0, 0, 0},
+	}
+
+	htBuf := new(bytes.Buffer)
+	err = binary.Write(htBuf, binary.LittleEndian, ht)
+	if err != nil {
+		return []byte{}, err
+	}
+	return htBuf.Bytes(), nil
+}
+
+func calculateSevLaunchDigest(firmwarePath, kernelPath, initrdPath, cmdline string) (res [sha256.Size]byte, err error) {
+	f, err := os.Open(firmwarePath)
+	if err != nil {
+		return res, err
+	}
+	defer f.Close()
+
+	digest := sha256.New()
+	if _, err := io.Copy(digest, f); err != nil {
+		return res, err
+	}
+
+	ht, err := constructSevHashesTable(kernelPath, initrdPath, cmdline)
+	if err != nil {
+		return res, err
+	}
+	digest.Write(ht)
+
+	copy(res[:], digest.Sum(nil))
+	return res, nil
+}
+
```

Comment:
Could we please move this set of SEV pre-attestation specific utilities into their own package? I would not mind seeing it live under `runtime/pkg/sev/utils.go` for example.